### DISCUSSION TITLE
Optimized Bazel remote caching - 2021 check in [CDNs, FUSE, disk_cache, etc]

### DISCUSSION THREAD

**Discussion by jerrymarino** [MobileNativeFoundation/discussions#61] [General]
Good afternoon,

First off, thanks to Keith and everyone else for getting this board up and running 👏 

I wanted to get a pulse on optimal remote caching in 2021 for Bazel and enumerate the problem space in general.

I'd like to know the state of the art on mechanisms you're using to improve efficiency. For example:
- [Build without the bytes](https://github.com/bazelbuild/bazel/issues/6862) has the ability to minimize downloads given a persistent cache during builds, are you using this?
- Solving this at the file system level: for instance giving Bazel a `disk_cache` that lazily fetches or eagerly fetches material
- Mitigating transfer cost inside of Bazel: for instance providing a custom file system module to noop downloads.
- Techniques to optimize outputs in the remote cache, for instance [Index While Building V2](https://github.com/bazel-ios/rules_ios/blob/master/docs/index_while_building.md)

I'm additionally curious about the infrastructure you've found optimal when hitting the cache on the road or OOO, as this poses an interesting challenge. For instance:
- fronting a cache with a secured CDN, like AWS or Google
- Accessing material over VPN and alternative security models e.g. HTTPs auth

Finally, I'm highly interested for those who have turned off remote caching and would love to hear about that too! Bazel is overall awesome even without an RBE or cache IMO.

**Comment by zachgrayio** [comment]
This is a big area of focus for us as you might imagine, so I can offer some comments.

1: build without the bytes

This one is pretty important and to non-bazelers its probably useful to quickly explain. By default, bazel currently will pull all of the intermediate artifacts down from the remote cache, resulting in what might be considered "unnecessary" load on the shared caching infrastructure as well as greatly decreased build performance, as there's so much additional work to be done. This is actually a bit different than the default behavior of other similar systems like Buck, and it can be somewhat surprising to users. 

Thus, "build without the bytes" essentially means "download only what is necessary" where "necessary" is determined by the action graph of the command & pattern you've invoked and the flags you've supplied to bazel governing the remote download level etc. 

There's certainly still a few rough edges here, but **making use of these controls is absolutely necessary to optimize your build**. It's one of the first things we look for when auditing the health of a bazel deployment in the wild, and it's hard to exaggerate how much of an impact setting these flags will have on large builds if they've been overlooked (70-80% end to end runtime reduction in some cases where network is the bottleneck). 

One of the blockers folks run into here is that getting this behavior correct in a remote cache/exec implementation brings with it a few challenges: Primarily, Bazel needs the outputs that it _expects_ to find remotely to _actually_ exist, and if they do not, it will crash with a fatal error as it's unable to recover from missing artifacts (this would require invalidation of and re-evaluation of large portions of the graph, something it currently is not designed to do). Squaring this behavior with the need to implement something like an LRU or TTL-based "garbage collection" mechanism to evict stale content & prevent the artifact cache storage from simply growing infinitely (quite costly at scale) can be a challenge, and as such there's.... at best, "varying" levels of support among the open-source examples of these APIs.

There's many other downsides here as well but in general the perf win and is usually worth the trouble and tradeoffs of getting this enabled for a given project. 

2: Solving this at the file system level: for instance giving Bazel a disk_cache that lazily fetches or eagerly fetches material

The current implementation of the disk_cache is fairly minimal and not necessarily performance optimized; we observed one instance of our cache storage engine locally outperforming disk_cache within the first few months of performance tuning it which I found surprising, but I never looked too deep into root cause there. To me this one sounds like a bit of a dead end and we never pursued it, but:

Filesystem-based solutions are definitely interesting and can offer a novel solution to this issue. There's some effort from some folks in the community on this front and we're definitely interested to see where that leads, but it's probably a long way off. FWIW, we're doing some things in this area internally, but that's all I can share.

3: fronting a cache with a secured CDN, like AWS or Google

Love this question. IMO, It's **only a matter of time before some sort of distributed/cdn functionality is uniformly offered by every solution** out there, the pitfalls of a centralized storage solution are fairly obvious here.

"traditional" CDNs aren't well suited to this problem for a few reasons, some of which touched on above, but in general, this _is_ an important piece of the infrastructure at scale, especially in an increasingly fully-remote world where team members need to access the shared cache (typically centralized in a cluster running in one AZ) from varying geographies. I would be skeptical of making use of any http cdn here for a few reasons, but also curious if anyone has tried it. 

We've actually gone to the lengths of creating a "CDN for bazel" which intelligently caches the "hottest" actions/blobs at the edge and we put nodes as close as we can to CI infrastructure and dev teams, and with huge success: getting the latency as close to 0ms as possible has a pretty significant impact for a few reasons. Firstly because gRPC has always had trouble with throughput when latency is a factor (this has improved over time but it still isn't perfect) and secondly, because Bazel itself just needs to make so many calls to the remote apis during a build, the overhead from a bit of latency really adds up. There's actually a recent PR to Bazel open to optimize some of that also.

4: Accessing material over VPN and alternative security models e.g. HTTPs auth

Have seen some folks leaving caches "wide open" and just accessing over the VPN... Obviously don't really recommend this for a few reasons for local builds, but I have seen it.

Other good options for security definitely exist and work well, I'm sure you're aware of some of the flags in bazel to include custom headers; can supply API keys this way or `Authorization` headers with bearer tokens (that's what `--google_credentials` does under the hood). Also, mutual tls is supported as well nowadays. 

5: Techniques to optimize outputs in the remote cache

This is kinda broad, but brings one thing to mind -- the version of the remote apis bazel implements doesn't support compression, though this is something actively discussed in the remote apis more broadly with support seeming imminent. However, the storage engine we wrote for the persistent layer of our cache and exec solution does optionally support compression, and the resultant compression ratios we observed on average across our workloads was really rather impressive. SSDs ain't cheap, so we were happy to observe this, and looking forward to seeing compression supported more widely by the APIs and clients.

**Reply by jerrymarino** [reply]
@zachgrayio thanks for taking the time out of your day to explain all of this, and great to hear about all the cool stuff you're up to at flare!

> This one is pretty important and to non-bazelers its probably useful to quickly explain.

I think you've put it very well. Bazel basically replaces the compiler invocations with a network request for the same output, and has gotten significantly better over the years. Still, given the low level integration, it needs to fetch a _ton_ of intermediates - which BWOB is an attempt to mitigate!

I'd be curious to know your anecdotes from large c++ bazel builds using this. If you're not caching link actions, does that mean in practice it you'd still need to fetch all object files again with BWOB? Or do you see a benefit to caching the archive actions along with the objects and debug information? If you cache static archives it basically amounts to batch downloading.

I am curious to see how the final outcome of BWOB stacks up against the other known techniques to mitigate over fetching in Bazel. e.g. as people discussed mentioned this [REAPI thread](https://groups.google.com/g/remote-execution-apis/c/qOSWWwBLPzo/m/TbcNYzwMBwAJ).

> We've actually gone to the lengths of creating a "CDN for bazel" which intelligently caches the "hottest" actions/blobs at the edge and we put nodes as close as we can to CI infrastructure and dev teams, and with huge success: getting the latency as close to 0ms as possible has a pretty significant impact for a few reasons. 

Totally agree: end to end cache latency usually ends up being a make or break factor in any remote cache. By CDN are you running on a shared CDN like cloudflare, aws, or running your own racks?

> Filesystem-based solutions are definitely interesting and can offer a novel solution to this issue. 

Yep, perhaps it'd be interesting a program to have a program that "warm" a `disk_cache` out of band e.g. git checkout events. Then point bazel build only at the file system disk cache. This way, people could get some benefits of the remote cache without waiting on a network request on every clang invocation. I haven't tested it out much yet, I'd be curious to hear more.

**Reply by rescuerunaways** [reply]
Following up on this: @zachgrayio has recently put together a landing page for Flare CDN solution, if folks reading this are interested to check it out: [https://flare.build/products/flare-cdn](https://flare.build/products/flare-cdn)

**Comment by brentleyjones** [comment]
1. **Remote Build without the Bytes:** We are in the process of getting Remote Execution setup for iOS. As part of that we are re-evaluating being able to use BwoB (with both remote-exec and remote-cache). Out of the box neither `--remote_download_minimal` or `--remote_download_toplevel` work super well for what we use it for (integrating with Xcode), since it won't download swiftmodules and indexstores. We've created a custom rule that uses aspects under the hood to collect all of the outputs we care about and re-exposes them, which allows `--remote_download_toplevel` to properly download them.
2. **Smart Disk Cache:** Remote Execution doesn't support using `--disk_cache`, so we had to explore what our options were here. So far we plan on using [bb-storage](https://github.com/buildbarn/bb-storage) or [bb-clientd](https://github.com/buildbarn/bb-clientd) as a local proxy to remote-exec/remote-cache. This will give us a size-managed disk cache (with a lot of ways we can customize) that is smart in other ways as well: existence caching to reduce network calls and the potential for lazy-loading results using it as a Remote Output Service (though FUSE performance currently isn't good enough for this on macOS afaict).
3. **Mitigating transfer cost inside of Bazel:** We haven't done anything in this space yet.
4. **Techniques to optimize outputs in the remote cache:** We plan on using Index While Building V2 when it lands. Bandwidth is a big issue with remote caching, and reducing what needs to be downloaded is a huge win.
5. **Non-office remote-cache usage:** We trying to optimize for the WFH/low bandwidth cases. We don't have our cache (and in the future, remote-exec) behind VPN, and instead use other auth mechanisms. We are looking into having cache nodes that are on edges closer to devs, but that's a bit out in our plans.

**Reply by zachgrayio** [reply]
>We've created a custom rule that uses aspects under the hood to collect all of the outputs we care about and re-exposes them, which allows --remote_download_toplevel to properly download them.

yeah unfortunately there's a lot of instances like this at the moment, but the juice often _is_ worth the squeeze. another pain point is the coverage tooling which raises actually quite a few issues, both with the runfiles the coverage merger needs and fails to find and also in the actual downloading of coverage artifacts... I'm guilty of some pretty ugly workarounds here.

is there anything you're considering open-sourcing on the aspect referenced above?

**Reply by jerrymarino** [reply]
> We are in the process of getting Remote Execution setup for iOS

Are you all planning to use buildbarn for this or something else? I saw the later comment about this so assumed so. Thanks for sharing all of the details of this and wish you the best luck on the project!

> download swiftmodules and indexstores. 

Maybe there's room for such an aspect in `rules_swift` or `rules_ios` if it's possible to generalize it enough for any users of BWOB. Could be useful for me too.

> Smart Disk Cache: Remote Execution doesn't support using --disk_cache, so we had to explore what our options were here. So far we plan on using bb-storage or bb-clientd as a local proxy to remote-exec/remote-cache. 

Awesome thanks so much for sharing that out, I haven't worked much with the buildbarn stack yet and I'm now interested to see how this piece can be used. Specifically, I wonder if one step towards remote exec while setting up remote cache would be getting the object sync setup with bb-storage. A while back, you could use the `remote_executor` flag and 1.0 GRPC interface as a remote cache without any execution as sort of a "stepping stone".  🤔

**Reply by brentleyjones** [reply]
@zachgrayio 
> is there anything you're considering open-sourcing on the aspect referenced above?

After we are good with it, yeah, anything that makes sense. It's pretty tailored to our setup, but at a minimum we can write up about the guts so others can make their own, if ours isn't generalizable enough.

**Reply by brentleyjones** [reply]
@jerrymarino
>Are you all planning to use buildbarn for this or something else?

We are evaluating a couple different options right now. BuildBarn has good tech to use on the client side though, thus the comments about that.

>Maybe there's room for such an aspect in rules_swift or rules_ios if it's possible to generalize it enough for any users of BWOB. Could be useful for me too.

Yeah, we'll either upstream or write about it!
